examples/tutorial_3_streamproto.md

```diff
@@ -0,0 +1,98 @@
+In the [previous tutorial](tutorial_2_customproto.md), we dug a little deeper in to how the ping protocol works by creating a custom ping protocol ourselves.
+
+This tutorial will focus on how to create a custom streaming protocol using libp2p.
+
+# Custom protocol in libp2p
+Let's create a `part3.nim`, and import our dependencies:
+```nim
+import bearssl
+import chronos
+import stew/byteutils
+
+import libp2p
+```
+This is similar to the second tutorial. Next, we'll declare our custom protocol.
+```nim
+const TestStreamCodec = "/test/stream-proto/1.0.0"
+
+type TestStreamProto = ref object of LPProtocol
+```
+
+Just as we did in the last tutorial, we've set a [protocol ID](https://docs.libp2p.io/concepts/protocols/#protocol-ids), and created a custom `LPProtocol`.
+
+As in the last tutorial, we are going to handle our client and server parts separately. The server will wait for an instruction from the client, then send a stream of data to the client and finish by closing the connection. The client will listen in a loop until the server has finished and closed the connection. Then the client will close its end of the connection as well.
+
+Lets start with the server part:
+```nim
+proc new(T: typedesc[TestStreamProto]): T =
+  # every incoming connections will in be handled in this closure
+  proc handle(stream: Connection, proto: string) {.async, gcsafe.} =
+    echo "[server] received from client: ", string.fromBytes(await stream.readLp(1024))
+
+    # send stream of data
+    for i in 0..10:
+      await stream.writeLp("Part " & $i)
+
+    # We must close the connections ourselves when we're done with it
+    await stream.close()
+
+  return T(codecs: @[TestStreamCodec], handler: handle)
+```
+Again, `handle`, will be called for each incoming peer asking for this protocol. In our simple example, the client sends `"please send data"` and once received, the server echos the message and sends a stream of data back. In this example, we are simply sending a series of strings as the stream of data, but this can be any data in a practical application.
+
+We can now create our client part:
+```nim
+proc streamData(p: TestStreamProto, stream: Connection) {.async.} =
+  await stream.writeLp("please send data")
+
+  # Read loop
+  while stream.atEof.not:
+
+    var buf = newSeq[byte](1024)
+    # readOnce reads length prefixed bytes into a buffer and returns the length
+    # of data read
+    let len = await stream.readOnce(addr buf[0], 1024)
```

Comment:
Don't use `readOnce` with `readLp/writeLp`. There are several reasons for that, one is that it breaks the symmetry of the two calls - there is no point in sending length prefixed data if at the end of the day you'll always read to `EOF` either way. The second is that you now cannot stream different types of messages over the same channel, since you've just made `EOF` your message marker. If the idea is to stream data, you should probably simply call `write` which writes the message as is, without length prefix.

Reply:
@dryajov I've removed `readLp/writeLp`, and instead am using `write` and `readOnce` only. Wdyt?

Reply:
removed `readLp/writeLp`, and instead am using `write`, `readOnce`, and `atEof`

Reply:
LGTM

Reply:
> I think 90% of users will use the Lp variants, but since we already show it in tutorial 2, it may make sense to show the non-Lp here But it should be clear that Lp can of be used for streaming as well, since it's about what everybody does

I'll make tutorial #3 a streaming tutorial with `readLp/writeLp` and `LPStreamEOFError`, then we can make a fourth with `read/write` and `atEof`.